src/containers/TransactionPage/TransactionPage.duck.js

```diff
@@ -396,6 +433,56 @@ export const declineSale = id => (dispatch, getState, sdk) => {
     });
 };
 
+export const cancelBeforeAccept = id => (dispatch, getState, sdk) => {
+  const curState = getState();
+  if (curState.TransactionPage.cancelBeforeAcceptInProgress) {
+    return Promise.reject(new Error('Transaction already in progress'));
+  }
+  dispatch(cancelBeforeAcceptRequest());
+
+  return sdk.transactions
+    .transition({ id, transition: TRANSITION_CANCEL_BOOKING_BEFORE_ACCEPT, params: {} }, { expand: true })
+    .then(response => {
+      dispatch(addMarketplaceEntities(response));
+      dispatch(cancelBeforeAcceptSuccess());
+      dispatch(fetchCurrentUserNotifications());
+      return response;
+    })
+    .catch(e => {
+      dispatch(cancelBeforeAcceptError(storableError(e)));
+      log.error(e, 'cancel-before-accept-failed', {
+        txId: id,
+        transition: TRANSITION_CANCEL_BOOKING_BEFORE_ACCEPT,
+      });
+      throw e;
+    });
+}
+
+export const cancelSale = (id, isCustomer) => (dispatch, getState, sdk) => {
+  const curState = getState();
+  if (curState.TransactionPage.cancelInProgress) {
+    return Promise.reject(new Error('Transaction already in progress'));
+  }
+  dispatch(cancelSaleRequest());
+
+  return sdk.transactions
+    .transition({ id, transition: isCustomer ? TRANSITION_CANCEL_BY_CUSTOMER : TRANSITION_CANCEL_BY_PROVIDER, params: {} }, { expand: true })
+    .then(response => {
+      dispatch(addMarketplaceEntities(response));
+      dispatch(cancelSaleSuccess());
+      dispatch(fetchCurrentUserNotifications());
+      return response;
+    })
+    .catch(e => {
+      dispatch(cancelSaleError(storableError(e)));
+      log.error(e, 'cancel-sale-failed', {
+        txId: id,
+        transition: isCustomer ? TRANSITION_CANCEL_BY_CUSTOMER : TRANSITION_CANCEL_BY_PROVIDER,
+      });
+      throw e;
+    });
+}
+
```

Comment:
That is two types of cancellation function. The first one is canceled before booking accepted by the provider that only runs by the customer, and the second cancel function is canceled after the booking accepted. So we need to split 2 types of cancelation functions.

Reply:
But on the look of it, the function's logic is almost the same, it only has transition difference. So I'm not convinced that we should duplicate the code out like this. If the only difference is the `transition` then it would make more sense to add a function to find the correct transition than duplicate the code's logic.